File: backend/supabase_client.py
import os
from typing import Optional
from dotenv import load_dotenv
from supabase import create_client, Client

load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    raise ValueError("Supabase credentials not found in .env")

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

def save_review(
    code: str,
    review: str,
    language: str,
    user_id: Optional[str] = None,
    repo_url: Optional[str] = None
):
    try:
        data = {
            "code": code,
            "review": review,
            "language": language,
            "user_id": user_id
        }
        if repo_url:
            data["repo_url"] = repo_url

        response = supabase.table("reviews").insert(data).execute()
        return response

    except Exception as e:
        print("Supabase save error:", str(e))
        raise


def get_user_reviews(user_id: str):
    try:
        response = supabase.table("reviews") \
            .select("*") \
            .eq("user_id", user_id) \
            .order("created_at", desc=True) \
            .execute()
        return response.data

    except Exception as e:
        print("Error fetching reviews:", str(e))
        return []


def get_cached_repo_review(repo_url: str, user_id: Optional[str] = None):
    """
    Check if a GitHub repo has already been reviewed using the dedicated repo_url column.
    Returns the most recent matching review, or None if no cache hit.
    """
    try:
        query = supabase.table("reviews") \
            .select("*") \
            .eq("language", "github-repo") \
            .eq("repo_url", repo_url) \
            .order("created_at", desc=True) \
            .limit(1)

        if user_id:
            query = query.eq("user_id", user_id)

        response = query.execute()
        return response.data[0] if response.data else None

    except Exception as e:
        print("Cache lookup error:", str(e))
        return None

File: backend/supabaseClient.js
// supabaseClient.js — Supabase persistence layer

const { createClient } = require("@supabase/supabase-js");
const config = require("./config");

let supabase = null;
if (config.supabaseUrl && config.supabaseKey) {
  supabase = createClient(config.supabaseUrl, config.supabaseKey);
} else {
  console.warn("⚠  Supabase credentials not set — database features will be unavailable.");
}

function requireSupabase() {
  if (!supabase) throw new Error("Supabase is not configured. Set SUPABASE_URL and SUPABASE_KEY in .env.");
  return supabase;
}

/**
 * Inserts a review record into the "reviews" table.
 * Throws on database error.
 */
async function saveReview({ code, review, language, userId = null, repoUrl = null }) {
  const row = { code, review, language, user_id: userId };
  if (repoUrl) row.repo_url = repoUrl;

  const { data, error } = await requireSupabase().from("reviews").insert(row).select();
  if (error) throw error;
  return data;
}

/**
 * Fetches a paginated list of reviews for a user, most recent first.
 * Returns { reviews, total, page, limit, totalPages }.
 */
async function getUserReviews(userId, { limit = 20, page = 1 } = {}) {
  const from = (page - 1) * limit;
  const to = from + limit - 1;

  const { data, error, count } = await requireSupabase()
    .from("reviews")
    .select("*", { count: "exact" })
    .eq("user_id", userId)
    .order("created_at", { ascending: false })
    .range(from, to);

  if (error) throw error;
  return {
    reviews: data || [],
    total: count ?? 0,
    page,
    limit,
    totalPages: Math.ceil((count ?? 0) / limit),
  };
}

/**
 * Returns the most recent cached review for a GitHub repo URL, or null if none.
 */
async function getCachedRepoReview(repoUrl, userId = null) {
  let query = requireSupabase()
    .from("reviews")
    .select("*")
    .eq("language", "github-repo")
    .eq("repo_url", repoUrl)
    .order("created_at", { ascending: false })
    .limit(1);

  if (userId) query = query.eq("user_id", userId);

  const { data, error } = await query;
  if (error) throw error;
  return data?.[0] ?? null;
}

module.exports = { saveReview, getUserReviews, getCachedRepoReview };
